### ISSUE TITLE
needless_borrow emitted for Iterator::take

### ISSUE THREAD

**Issue by willcrichton** [rust-lang/rust-clippy#8475]
### Summary

Iterator::take moves its receiver, but clippy believes that rustc would automatically mutably borrow its receiver.

### Lint Name

needless_borrow

### Reproducer

I tried this code:

```rust
use std::mem::drop;
fn main() {
  let mut it = [0, 1].into_iter();
  (&mut it).take(1).for_each(drop);
  (&mut it).take(1).for_each(drop);
}
```

I saw this happen:

```
warning: this expression borrows a value the compiler would automatically borrow
 --> src/main.rs:4:3
  |
4 |   (&mut it).take(1).for_each(drop);
  |   ^^^^^^^^^ help: change this to: `it`
  |
  = note: `#[warn(clippy::needless_borrow)]` on by default
```

I expected not to get a warning.

### Version

```text
rustc 1.60.0-nightly (30b3f35c4 2022-02-17)
binary: rustc
commit-hash: 30b3f35c420694a4f24e5a4df00f06073f4f3a37
commit-date: 2022-02-17
host: x86_64-apple-darwin
release: 1.60.0-nightly
LLVM version: 14.0.0
```

**Comment by Jarcho** [comment]
This should already be fixed in the latest nightly.

**Comment by willcrichton** [comment]
Yes, I double-checked and it is fixed on nightly. Sorry for the false false positive!
